File: lib/common/texture.dart
import 'dart:ffi';
import 'dart:io';

import 'package:ffi/ffi.dart';
import 'package:image/image.dart';
import 'package:nebula/nebula.dart';

class Texture {
  late final int handle;
  late String type;

  static Map<String, Texture> textureCache = <String, Texture>{};

  Texture(this.handle, this.type);

  static Texture loadFromFile(String path, String type) {
    if (textureCache.containsKey(path)) return Texture(textureCache[path]!.handle, type);

    int handle = 0;
    using((arena) {
      final textures = arena<UnsignedInt>();
      gl.glGenTextures(1, textures);
      handle = textures.elementAt(0).value;
    });

    gl.glActiveTexture(GL_TEXTURE0);
    gl.glBindTexture(GL_TEXTURE_2D, handle);

    final image = decodeImage(File(path).readAsBytesSync())!;
    using((arena) => gl.glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, image.width, image.height, 0, GL_RGBA, GL_UNSIGNED_BYTE, (arena<Uint8>(image.length)..asTypedList(image.length).setAll(0, image.getBytes())).cast()));

    gl.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
    gl.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);

    gl.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_REPEAT);
    gl.glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_REPEAT);

    gl.glGenerateMipmap(GL_TEXTURE_2D);

    final texture = Texture(handle, type);
    textureCache[path] = texture;
    return texture;
  }

  static Texture getDefaultAlbedo() => Texture.loadFromFile('Resources/default/default_albedo.png', 'albedo');
  static Texture getDefaultNormal() => Texture.loadFromFile('Resources/default/default_normal.png', 'normal');
  static Texture getDefaultMetallic() => Texture.loadFromFile('Resources/default/default_metallic_roughness.png', 'metallic');
  static Texture getDefaultRoughness() => Texture.loadFromFile('Resources/default/default_metallic_roughness.png', 'roughness');
  static Texture getDefaultAO() => Texture.loadFromFile('Resources/default/default_ao.png', 'ao');

  void bind(int unit) {
    gl.glActiveTexture(unit);
    gl.glBindTexture(GL_TEXTURE_2D, handle);
  }
}

class DefaultTextures {
  Texture? albedo;
  Texture? normal;
  Texture? metallic;
  Texture? roughness;
  Texture? ao;

  DefaultTextures({this.albedo, this.normal, this.metallic, this.roughness, this.ao});

  Texture getAlbedoOrDefault() => albedo ?? Texture.getDefaultAlbedo();
  Texture getNormalOrDefault() => normal ?? Texture.getDefaultNormal();
  Texture getMetallicOrDefault() => metallic ?? Texture.getDefaultMetallic();
  Texture getRoughnessOrDefault() => roughness ?? Texture.getDefaultRoughness();
  Texture getAOOrDefault() => ao ?? Texture.getDefaultAO();
}

File: src/Common/Texture.cs
using System.Collections.Generic;
using System.IO;
using OpenTK.Graphics.OpenGL4;
using StbImageSharp;

namespace Nebula;

public class Texture {
    public readonly int handle;
    public string type;

    public static Dictionary<string, Texture> textureCache = new Dictionary<string, Texture>();

    public static Texture LoadFromFile(string path, string type) { // = "default") {
        if (textureCache.ContainsKey(path)) return new Texture(textureCache[path].handle, type);

        int handle = GL.GenTexture();

        GL.ActiveTexture(TextureUnit.Texture0);
        GL.BindTexture(TextureTarget.Texture2D, handle);

        StbImage.stbi_set_flip_vertically_on_load(1);

        using (Stream stream = File.OpenRead(path)) {
            ImageResult image = ImageResult.FromStream(stream, ColorComponents.RedGreenBlueAlpha);
            GL.TexImage2D(TextureTarget.Texture2D, 0, PixelInternalFormat.Rgba, image.Width, image.Height, 0, PixelFormat.Rgba, PixelType.UnsignedByte, image.Data);
        }

        GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureMinFilter, (int)TextureMinFilter.Linear);
        GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureMagFilter, (int)TextureMagFilter.Linear);

        GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureWrapS, (int)TextureWrapMode.Repeat);
        GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureWrapT, (int)TextureWrapMode.Repeat);

        GL.GenerateMipmap(GenerateMipmapTarget.Texture2D);

        Texture texture = new Texture(handle, type);
        textureCache.Add(path, texture);
        return texture;
    }

    public Texture(int handle, string type)
    {
        this.handle = handle;
        this.type = type;
    }

    public static Texture GetDefaultAlbedo() => Texture.LoadFromFile("Resources/default/default_albedo_ao.png", "albedo");
    public static Texture GetDefaultNormal() => Texture.LoadFromFile("Resources/default/default_normal.png", "normal");
    public static Texture GetDefaultMetallic() => Texture.LoadFromFile("Resources/default/default_metallic_roughness.png", "metallic");
    public static Texture GetDefaultRoughness() => Texture.LoadFromFile("Resources/default/default_metallic_roughness.png", "roughness");
    public static Texture GetDefaultAO() => Texture.LoadFromFile("Resources/default/default_albedo_ao.png", "ao");

    public void Bind(TextureUnit unit)
    {
        GL.ActiveTexture(unit);
        GL.BindTexture(TextureTarget.Texture2D, handle);
    }
}

public class DefaultTextures {
    public Texture albedo;
    public Texture normal;
    public Texture metallic;
    public Texture roughness;
    public Texture ao;

    public DefaultTextures(Texture albedo = null, Texture normal = null, Texture metallic = null, Texture roughness = null, Texture ao = null)
    {
        this.albedo = albedo;
        this.normal = normal;
        this.metallic = metallic;
        this.roughness = roughness;
        this.ao = ao;
    }

    public Texture GetAlbedoOrDefault() => albedo != null ? albedo : Texture.GetDefaultAlbedo();
    public Texture GetNormalOrDefault() => normal != null ? normal : Texture.GetDefaultNormal();
    public Texture GetMetallicOrDefault() => metallic != null ? metallic : Texture.GetDefaultMetallic();
    public Texture GetRoughnessOrDefault() => roughness != null ? roughness : Texture.GetDefaultRoughness();
    public Texture GetAOOrDefault() => ao != null ? ao : Texture.GetDefaultAO();
}